[libc++][test] Don't include `test_format_context.h` in `parse.pass.cpp`

frederick-vs-ja opened the pull request:
The `parse.pass.cpp` tests doen't need to call `test_format_context_create` to create a `basic_format_context`, so they shouldn't include `test_format_context.h`.

The `to_address` mechanism works around the iterator debugging mechanisms of MSVC STL. Related to [LWG3989](https://cplusplus.github.io/LWG/issue3989).

Discovered when implementing `formatter<tuple>` in MSVC STL. With the inclusion removed, `std/utilities/format/format.tuple/parse.pass.cpp` when using enhanced MSVC STL (and `/utf-8` option for MSVC).

frederick-vs-ja committed:
[libc++][test] Don't include `test_format_context.h` in `format.tuple/parse.pass.cpp`

The test doesn't need to call `test_format_context_create` to create a `basic_format_context`, so it shouldn't include `test_format_context.h`.

Discovered when implementing `formatter<tuple>` in MSVC STL. With the inclusion removed, local test passes when using enhanced MSVC STL.

frederick-vs-ja committed:
Workaround for MSVC STL and LWG3989

mordante commented on libcxx/test/std/utilities/format/format.tuple/parse.pass.cpp:
```diff
@@ -41,7 +41,7 @@ constexpr void test(StringViewT fmt, std::size_t offset) {
 
   std::same_as<typename StringViewT::iterator> auto it = formatter.parse(parse_ctx);
-  assert(it == fmt.end() - offset);
+  assert(std::to_address(it) == std::to_address(fmt.end()) - offset);
```
Since this works for libc++ can you add a comment referring to the LWG issue you created.
(That should avoid people trying to "simplify the code.)
Searching for the original `assert` in the libc++ tests finds more occurrences. Can you fix these too?

mordante replied:
Can you also add a comment? Other than that if LGTM!

mordante replied:
yes please. I really want to discourage people in five years to "clean up" this code and break MSVC STL.

frederick-vs-ja committed:
Remove other redundant `test_format_context.h` inclusion

And consistently use
`assert(std::to_address(it) == std::to_address(fmt.end()) - offset)`
due to LWG3989

frederick-vs-ja committed:
Complete inclusion of `<memory>` and comments for workaround

frederick-vs-ja committed:
Merge branch 'main' into redux-test_format_context.h

mordante reviewed (approved):
Thanks! LGTM!

ldionne commented:
Merging. The CI failures are flukes.
